### ISSUE TITLE
TestNG should provide an Api which allow to find all dependent of a specific test

### ISSUE THREAD

**Issue by juherr** [cbeust/testng#893]
https://issues.apache.org/jira/browse/SUREFIRE-1073

**Comment by krmahadevan** [comment]
@juherr - What is the expectations of this issue. I would specifically like to know the following:

1. How will the user invoke this? Should this be made available only when someone creates a `TestNG` instance ?
2. What would the input for this api look like ?

It would be good if you could please help add some additional context into this issue.

**Comment by juherr** [comment]
It is for the integrations. The need is to add the dependent `TestMethod`s on `TestMethod` api

**Comment by krmahadevan** [comment]
You mean something like `public List<ITestNGMethod> getDownstreamDependencies()` into `ITestNGMethod` interface right ?

**Comment by juherr** [comment]
Yes, exactly. It exists in the model but not available by the api.

About the naming, I think something closer the annotation attribute name will be better.

**Comment by krmahadevan** [comment]
@juherr 

> About the naming, I think something closer the annotation attribute name will be better.

What attribute are you referring to here? In TestNG, we only have a mapping from a child (Dependent method) to a parent (Independent method upon which the dependent method depends on)

But here we are asking for all children, given a parent.

```java
public class TestClassSample {

  @Test
  public void independent() {}

  @Test(dependsOnMethods = "independent")
  public void dependent() {}

  @Test(dependsOnMethods = "independent")
  public void anotherDependent() {}
}
```

Here we are asking for an API, wherein if we give input as `independent()` it should return `dependent()` and `anotherDependent()`. Is this the right expectations?

If yes, then I think the api should be called as `getDownstreamDependencies()`. WDYT ? 

For upstream dependencies ( which is what `dependsOnMethods` represents) we have this `String[] getMethodsDependedUpon();` in the ITestNGMethod interface.

**Comment by juherr** [comment]
I will check again. I thought the need was to have getMethodsDependedUpon() returning object instead of string

**Comment by krmahadevan** [comment]
@juherr - I have raised a PR which now provides both upstream and downstream dependencies for a test method. Please check.
